### PR TITLE
[RuntimeObject] Allow grid based object to optimize collision checks

### PR DESCRIPTION
**Pull request by D8H**
This is built over:
* https://github.com/4ian/GDevelop/pull/3236

It adds a new method to get the hitboxes around a given area: `getHitBoxesAround(left: float, top: float, right: float, bottom: float)`
This method is used in collision checks so grid based objects like tile map or marching squares can be optimized.

The only impact on performances is on `separateFromObjects` when another object has more than 4 hitboxes. In this case, the AABB of `this` is calculated to try to get less hit boxes. Objects with a lot of hitboxes will probably be a tile map where optimizations can take place.

It will help to write a PR for this issue (the next thing I will do):
* https://github.com/4ian/GDevelop/issues/2775

### CODE REVIEW

File: GDJS/Runtime/runtimeobject.ts [hunk 1]
```diff
@@ -1640,40 +1684,134 @@ namespace gdjs {
...
+          for (const otherObject of otherObjects) {
+            if (otherObject.id === this.id) {
+              continue;
+            }
+            const otherHitBoxes = otherObject.getHitBoxesAround(
+              aabb.min[0],
+              aabb.min[1],
+              aabb.max[0],
+              aabb.max[1]
+            );
```
Comment (D8H) [inline GDJS/Runtime/runtimeobject.ts:1714]: The only impact should be in `SeparateFromObjects` where an AABB is asked on 1 object.
If the object has less than 4 hit boxes we could avoid to calculate its AABB and just call the "non optimized" method.

**Comment by D8H** [comment]
The performance impact should be nearly inexistent now.

**Review by 4ian — COMMENTED** [review]
Looks good I think from a quick check. Great stuff!! 😄

It's important to understand that `getHitboxesAround` is taking the area in the scene space coordinates. So objects like tilemap will have to consider any transformation they have (i.e: they must convert from the scene space coordinates to their local coordinates, to then find the hitboxes in the area) (unless I'm mistaken).

File: GDJS/Runtime/runtimeobject.ts [hunk 2]
```diff
@@ -1253,6 +1256,37 @@ namespace gdjs {
...
+     *
+     * This method can be override by grid based objects to optimize collision checks.
+     *
+     * When overriding this method, the following ones should be overridden too:
+     * * {@link getHitBoxes}
+     * * {@link getAABB}
+     * * {@link updateHitBoxes}
+     * * {@link updateAABB}
+     *
+     * @param left bound of the area
```
Comment (4ian) [inline GDJS/Runtime/runtimeobject.ts:1274]: Let's precise "left bound of the area, in scene coordinates" and for the 3 others too :)

**Comment by 4ian** [comment]
> This method is used in collision checks so grid based objects like tile map or marching squares can be optimized.

This is used in:
- platformer engine
- The separate action: `separateFromObjects`, `separateFromObjectsList` 
- The collision condition (`collisionTest`)
- The raycasting condition (`raycastTest`)
- The point condition/`isCollidingWithPoint`

Anything else you think we might have missed?

**Comment by D8H** [comment]
> > This method is used in collision checks so grid based objects like tile map or marching squares can be optimized.
> 
> This is used in:
> 
>     * platformer engine
> 
>     * The separate action: `separateFromObjects`, `separateFromObjectsList`
> 
>     * The collision condition (`collisionTest`)
> 
>     * The raycasting condition (`raycastTest`)
> 
>     * The point condition/`isCollidingWithPoint`
> 
> 
> Anything else you think we might have missed?

In `runtimescene-pixi-renderer`, the call is to draw hitboxes for debugging which is ok.

```
davy@PC2014-06:/mnt/c/WorkspaceGDevelop/GDevelop/GDJS/Runtime$ grep -rn "getHitBoxes(" .
./pixi-renderers/runtimescene-pixi-renderer.ts:212:        const hitboxes = object.getHitBoxes();
./runtimeobject.ts:1246:    getHitBoxes(): gdjs.Polygon[] {
./runtimeobject.ts:1287:      return this.getHitBoxes();
./runtimeobject.ts:1646:      const hitBoxes = this.getHitBoxes();
./runtimeobject.ts:1654:        let otherHitBoxesArray = otherObject.getHitBoxes();
./runtimeobject.ts:1706:      const hitBoxes = this.getHitBoxes();
./runtimeobject.ts:1718:            let otherHitBoxesArray = otherObject.getHitBoxes();
davy@PC2014-06:/mnt/c/WorkspaceGDevelop/GDevelop/GDJS/Runtime$
```

I didn't change the lighting extension because I don't know it enough, but I could take a look. 
In `platformerobjectruntimebehavior`, the call is on the character which is legit. If a character is a tilemap, it should be small enough.

```
davy@PC2014-06:/mnt/c/WorkspaceGDevelop/GDevelop/Extensions$ grep -rn "getHitBoxes(" .
./Lighting/lightruntimeobject-pixi-renderer.ts:60:          runtimeObject.getHitBoxes()[0].vertices[i]
./Lighting/lightruntimeobject-pixi-renderer.ts:389:      const lightHitboxPoly = this._object.getHitBoxes()[0];
./Lighting/lightruntimeobject-pixi-renderer.ts:399:        obstacleHitBoxes[i] = lightObstacles[i].behavior.owner.getHitBoxes();
./PlatformBehavior/platformerobjectruntimebehavior.ts:2045:      for (const hitBox of behavior.owner.getHitBoxes()) {
davy@PC2014-06:/mnt/c/WorkspaceGDevelop/GDevelop/Extensions$
```

There is no usage in the extension repository.
```
davy@PC2014-06:/mnt/c/WorkspaceGDevelop/GDevelop-extensions/Extensions$ grep -rn "getHitBoxes(" .
davy@PC2014-06:/mnt/c/WorkspaceGDevelop/GDevelop-extensions/Extensions$
```

Also, I wanted to ask for making `getHitBoxes` an `Iterable` to avoid to force grid based implementations to produce a flat array.

File: GDJS/Runtime/runtimeobject.ts [hunk 3]
```diff
@@ -1253,6 +1256,37 @@ namespace gdjs {
       return this.hitBoxes;
     }
 
+    /**
+     * Get the hit boxes for the object around a given area.
```
[the remark this answers is not in the document]
Comment (D8H) [reply GDJS/Runtime/runtimeobject.ts:1260]: The following lines explain it:
```
     * The hit boxes don't need to actually overlapping the area,
     * but the ones that do must be returned.
```
This is method can return every hit-boxes when there is no specific optimizations. And optimized implementations can give useless hit-boxes if it's a good trade of.

[the remark this answers is not in the document]
Comment (D8H) [reply GDJS/Runtime/runtimeobject.ts:1260]: It could be: "Return at least all the hit boxes overlapping a given area"

**Comment by D8H** [comment]
I made the changes for the lighting extension.
This extension implementation do very odd things and probably need optimizations. I added a few comments.

**Comment by D8H** [comment]
I fix a bug while testing it with a tile map platformer.

**Comment by 4ian** [comment]
Agree on having more tests once the tile map implementation is done (we could also have a GridHitBoxesTestRuntimeObject or something like this if we feel like we need a test object for this).

For the lighting thanks for adding comments where things are not optimal - the end of the implementation was maybe a bit too rushed when we wanted to get it done.

**Comment by 4ian** [comment]
@D8H Do you want to rebase this on master? :) As the previous PR is now merged and going in the new release of today, I think we're getting ready to get this in, so then the PR on tilemaps can progress :)

**Comment by 4ian** [comment]
Thank you for the rebase. I aim to review and merge this by Friday (it's essentially the lights where I need to double check your comments) so you are assured we have this in to do the tilemap stuff :)

File: Extensions/Lighting/lightruntimeobject-pixi-renderer.ts [hunk 4]
```diff
@@ -379,38 +387,54 @@ namespace gdjs {
...
         for (let j = 0; j < 2; j++) {
+          // This is actually completely useless as
+          // this._lightBoundingPoly.vertices[i] === lightHitboxPoly.vertices[i]
           this._lightBoundingPoly.vertices[i][j] =
             lightHitboxPoly.vertices[i][j];
         }
       }
       const obstaclesCount = lightObstacles.length;
-      const obstacleHitBoxes = new Array(obstaclesCount);
+      // This array could be avoided.
```
Comment (4ian) [inline Extensions/Lighting/lightruntimeobject-pixi-renderer.ts:412]: You're correct this intermediate array is not useful.

File: Extensions/Lighting/lightruntimeobject-pixi-renderer.ts [hunk 5]
```diff
@@ -430,6 +454,10 @@ namespace gdjs {
           }
         }
       }
+      // obstaclePolygons[0] === this._lightBoundingPoly
+      // this._lightBoundingPoly === this._object.hitBoxes
+      // This is actually modifying the object hit boxes.
+      // Is it intended?
```
Comment (4ian) [inline Extensions/Lighting/lightruntimeobject-pixi-renderer.ts:460]: I don't think this is intended.

The idea in the algorithm is to always have a rectangle which is at least as large as the light (i.e the rectangle which has a width/height of `2*radius` with as center the center of the light), but if obstacles are outside this rectangle we extend it.

But this should not modify the hitbox. So the renderer here should keep its own `this._lightBoundingPoly` without pointing to the vertices of the object hitbox - which is wrong (but thankfully not creating issues, so far at least)

**Review by 4ian — COMMENTED** [review]
So the issues in the light runtime object renderer boil down to:

- `_lightBoundingPoly` is, wrongly, a shallow copy of the hitbox, sharing the vertices. Instead we should change this so that it owns its own vertices.
  Note that`_lightBoundingPoly` is still useful I think. It's a quadrilateral at least as large as the light radius, and can be extended if obstacles are going beyond it.
- `obstacleHitBoxes` can be removed, because we can directly create `obstaclePolygons`.

I'll make the fixes tomorrow (unless you want to give it a try). Then if we've not forgot any call to `getHitBoxesAround`, I think we'll be good to go.

**Comment by 4ian** [comment]
I've hopefully fixed the wrong stuff for the light - not perfect yet I guess but better.

**Review by 4ian — APPROVED** [review]

**Comment by 4ian** [comment]
Looks good to go, merging this now! Thanks for working on this, looking forward to how this can be useful with the tilemap collisions :D

**Pull request merged**